keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,364 @@
+* Kubelet spans can detect node-level latency - each node produces trace data filterable with node hostname
+    * Are there issues with particular regions? Cloud providers? Machine-types?
+
+### Tracing Requests and Exporting Spans
+
+The gRPC client in the kubelet's Container Runtime Interface (CRI) [Remote Runtime Service](https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/kubelet/cri/remote/remote_runtime.go) and [the CRI streaming package](https://github.com/kubernetes/kubernetes/tree/release-1.21/pkg/kubelet/cri/streaming) will be instrumented to export and propagate trace data.
+The [Go implementation of OpenTelemetry](https://github.com/open-telemetry/opentelemetry-go) will be used.
+An [OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlptrace/otlptracegrpc/exporter.go) and
+an [OTLP trace provider](https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/provider.go) along with a [context propagator](https://opentelemetry.io/docs/go/instrumentation/#propagators-and-context) will be configured.
+
+Also, designated http servers and clients will be wrapped with [otelhttp](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/v0.21.0/instrumentation/net/http)
+to generate spans for sampled incoming requests and propagate context with client requests.
+
+OpenTelemetry-Go provides the [propagation package](https://github.com/open-telemetry/opentelemetry-go/blob/main/propagation/propagation.go) with which you can add custom key-value pairs known as [baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md). Baggage data will be propagated across services within contexts.
+
+### Running the OpenTelemetry Collector
+
+The [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector) can be run as a sidecar, a daemonset, a deployment , or a combination in which the daemonset buffers telemetry and forwards to the deployment for aggregation (e.g. tail-base sampling) and routing to a telemetry backend.  To support these various setups, the kubelet should be able to send traffic either to a local (on the control plane network) collector, or to a cluster service (on the cluster network).
+
+### Kubelet Configuration
+
+Refer to [OpenTelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/ccbfdd9020ba0d581aa5d880235302f8b0eb8669/specification/resource/semantic_conventions/README.md#service)
+
+```golang
+// KubeletConfiguration contains the configuration for the Kubelet
+type KubeletConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
+    ------------
+
+    // +optional
+    // TracingConfiguration
+    TracingConfiguration TracingConfiguration
+}
+
+// TracingConfiguration specifies configuration for opentelemetry tracing
+type TracingConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
```

Comment:
Is it normal to have TypeMeta embedded in a struct that is embedded in something that also has TypeMeta?

Reply:
Oh, I don't think so.  [LoggingConfiguration](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go#L386) & [MemorySwapConfiguration ](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go#L336)don't, and I can't find any that do. Thanks, again. Removed that.